lib/Sema/CSDiag.cpp

```diff
@@ -4555,6 +4555,188 @@ typeCheckArgumentChildIndependently(Expr *argExpr, Type argType,
+  auto &TC = CS->TC;
+  // Call expression is formed as 'foo.bar' where 'foo' might be an
+  // implicit "Self" reference, such use wouldn't provide good diagnostics
+  // for situations where instance members have equal names to functions in
+  // Swift Standard Library e.g. min/max.
+  auto UDE = dyn_cast<UnresolvedDotExpr>(fnExpr);
+  if (!UDE)
+    return false;
+
+  auto baseExpr = dyn_cast<DeclRefExpr>(UDE->getBase());
+  if (!baseExpr)
+    return false;
+
+  auto baseDecl = baseExpr->getDecl();
+  if (!baseExpr->isImplicit() || baseDecl->getName() != TC.Context.Id_self)
+    return false;
+
+  // Our base expression is an implicit 'self.' reference e.g.
+  //
+  // extension Sequence {
+  //   func test() -> Int {
+  //     return max(1, 2)
+  //   }
+  // }
+  //
+  // In this example the Sequence class already has two methods named 'max'
+  // none of which accept two arguments, but there is a function in
+  // Swift Standard Library called 'max' which does accept two arguments,
+  // so user might have called that by mistake without realizing that
+  // compiler would add implicit 'self.' prefix to the call of 'max'.
+  ExprCleaner cleanup(argExpr);
+
+  // Let's type check argument expression without any contextual information.
+  ConcreteDeclRef ref = nullptr;
+  auto typeResult = TC.getTypeOfExpressionWithoutApplying(argExpr, CS->DC, ref);
+  if (!typeResult.hasValue())
+    return false;
+
+  auto argType = typeResult.getValue();
+
+  // If argument type couldn't be properly resolved or has errors,
+  // we can't diagnose anything in here, it points to the different problem.
+  if (isUnresolvedOrTypeVarType(argType) || argType->hasError())
+    return false;
+
+  // Let's construct a list of the candidates from all upper contexts,
+  // match the argument expression against that and check if it does match.
+  llvm::SmallSet<LookupResult::Result, 4> choices;
+
+  auto context = CS->DC;
+  using CandididateMap =
+      llvm::SmallDenseMap<ValueDecl *, llvm::SmallVector<OverloadChoice, 2>>;
+
+  auto getBaseKind = [](ValueDecl *base) -> std::string {
+    if (!base)
+      return "module";
+
+    auto context = base->getDeclContext();
+    DescriptiveDeclKind kind = DescriptiveDeclKind::Module;
+
+    do {
+      if (isa<ExtensionDecl>(context))
+        return "extension of";
+
+      if (auto nominal = dyn_cast<NominalTypeDecl>(context)) {
+        kind = nominal->getDescriptiveKind();
+        break;
+      }
+
+      context = context->getParent();
+    } while (context);
+
+    return Decl::getDescriptiveKindName(kind);
+  };
+
+  std::function<std::string(DeclContext *)> getBaseName =
+      [&](DeclContext *context) -> std::string {
+    switch (context->getContextKind()) {
+    case swift::DeclContextKind::FileUnit:
+      return getBaseName(context->getParent());
+
+    case swift::DeclContextKind::AbstractFunctionDecl:
+      return cast<AbstractFunctionDecl>(context)->getNameStr();
```

Comment:
I don't think it's reachable, I think it can only be if we were to accept local functions. Thanks for the suggestion, @rintaro, I'll change modify the code accordingly.